### ISSUE TITLE
POJO AvroSchema always allowNull

### ISSUE THREAD

**Issue by sijie** [apache/pulsar#3741]
**Describe the bug**

Currently AvroSchema makes all fields allowNull, which makes the generated schema doesn't respect what POJO defines.

**Expected behavior**

AvroSchema should respect to what POJO defines

**Additional context**

N/A

**Comment by sijie** [comment]
@codelipenghui 's team is working on this issue.

**Comment by jerrypeng** [comment]
@sijie I don't quite follow.  If a class has a field "Integer" than the field shoulda allow vs if the field is "int" then it shouldn't:

```
 public static class Foo {
        int field1;
        Integer field2;
    }

        log.info("{}", new String(AvroSchema.of(Foo.class).getSchemaInfo().getSchema()));

```

produces:

```
{"type":"record","name":"Foo","namespace":"org.company.Test4$","fields":[{"name":"field1","type":"int"},{"name":"field2","type":["null","int"],"default":null}]}

```

which is correct.  Am I missing something?

**Comment by sijie** [comment]
@jerrypeng 

the main problem of this is : if we are serializing POJO into avro formats, it should be respecting to how AVRO generates schema for POJOs. Otherwise, if you are migrating other AVRO data from other data sources to Pulsar, or from Pulsar to other data destinations, it will raise schema incompatibility issue. Because Pulsar schema generated schema always allow null, while the schema generated by other systems don't allow null, even they are generated from same POJO. 

AVRO is a pretty standard data format. If pulsar is using AVRO schema, we should respect to its rules, otherwise it is impossible for pulsar to be used for exchanging AVRO data.

**Comment by jerrypeng** [comment]
@sijie I don't understand why you say we are not respecting the rules of AVRO, we are using the official java AVRO library to schema generation.  And if another system is using AVRO java library to generate POJOs (java objects) then they will get the same result.  Are you talking about cross language support? The only flag that could be settable is the allow null flag that gets set in Pulsar's Avro schema.  We can expose that flag to users.

**Comment by skyrocknroll** [comment]
@jerrypeng usually we create POJO by defining avro schema file and create it by using avro maven plugin. Once we provide the generated POJO to pulsar producer the schema registered in registry not identical to schema we have written. This is because of allowNull as pointed out by sijieg . This came as an surprise . When we tried consuming with Python with the same schema it failed because of schema incompatibility. If people want null to all the fields then they can add it at the avro schema.

**Comment by skyrocknroll** [comment]
Exposing the flag may also work with default as allowNull to false .

**Comment by jerrypeng** [comment]
@skyrocknroll @sijie I don't think the current API is the problem, I think we just need to separate API to specify a custom schema.  We currently only allow users to derive a schema from a Java class using the Avro library.  This generates an Avro schema that matches the semantics of java i.e. classes can be null and primitives are not.  However, if I have my own custom avro schema, I can arbitrarily set one class to be nullable while the another class is not.  There is not way for Avro to know that information from just a java class definition.  Even by allowing users to set the flag "allowNull" will not cover all cases as user can arbitrarily create avro schemas will null or not null fields.  I think we can do the following: 

1. Use  Avro java annotation to massage your java class to have the schema you desire
2. Have an API for users to specify the schema they want.

**Comment by merlimat** [comment]
Irrespective of the `allowNull=true/false` question, I think: 
 1. This seems to be a problem only with Avro generated code, in which there was a schema defined and we're extracting the wrong schema from the generated POJO
 2. This shouldn't be an issue when one defines an ad-hoc POJO (without starting from Avro schema), because annotations can be used to request nullable/non-nullable.

The easiest solution would be to detect that a POJO is generated by Avro and use the `getClassSchema()` instead of doing the reflection based approach.

**Comment by sijie** [comment]
> I don't understand why you say we are not respecting the rules of AVRO
> The only flag that could be settable is the allow null flag that gets set in Pulsar's Avro schema. We can expose that flag to users.

@jerrypeng 

That is the problem I am reporting this in the issue. Because Pulsar Avro Schema always set AllowNull and pulsar's users don't have the ability to bypass the AllowNull. My point here is Pulsar AVRO should not try to attach AllowNull.

> If people want null to all the fields then they can add it at the avro schema.

@skyrocknroll exactly.

> I don't think the current API is the problem, I think we just need to separate API to specify a custom schema. 

Jerry, that's a different feature. 

The point here for a given POJO, Pulsar should maintain a basic contract: the schema generated by Pulsar AVRO should be same/compatible with what other tools generated. Pulsar AVRO should not automatically attach AllowNull for all fields, otherwise we shouldn't call it AVRO schema, we should be calling it AlwaysAllowNullAvroSchema.

> This seems to be a problem only with Avro generated code, in which there was a schema defined and we're extracting the wrong schema from the generated POJO
> This shouldn't be an issue when one defines an ad-hoc POJO (without starting from Avro schema), because annotations can be used to request nullable/non-nullable.

@merlimat : that's not the *only* case. People can use ReflectData to generate a schema and write events in a different system (let's call it source). when people migrate data from source to pulsar, people use same POJO, these two schemas are not compatible, which will result in events are not able to be written to pulsar.

> The easiest solution would be to detect that a POJO is generated by Avro and use the getClassSchema() instead of doing the reflection based approach.

That's a very limited solution to cover only one use case. 

---

In summary, the questions is about "shall Pulsar automatically attach AllowNull to POJO generated schema". Attaching AllowNull changes the default behavior of how AVRO generates POJO and makes Pulsar inconsistent with how other systems use AVRO.

If the team decides that AllowNull should be the default behavior in AvroSchema, it should be well documented and highlighted to all avro schema users. and a separate AvroSchema or a flag to disable AllowNull should also be provided for users to not use AllowNull.

I don't think detecting if it is an Avro generated POJO is the right approach. We should just remove (or provide the ability to) AllowNull when constructing AvroSchema.

**Comment by merlimat** [comment]
> that's not the only case. People can use ReflectData to generate a schema and write events in a different system (let's call it source). when people migrate data from source to pulsar, people use same POJO, these two schemas are not compatible, which will result in events are not able to be written to pulsar.

Given a POJO generated by Avro, there is no way to determine whether this POJO was generated with a schema that allowed or not for null types.

> That's a very limited solution to cover only one use case.

It's a 100% correct solution for that case. I don't see what's limited about that.

**Comment by merlimat** [comment]
Though, I too think we should remove the AllowNull from the generator, but that's separate from the point that that alone won't guarantee we can generate the correct schema starting from the generated POJO.

**Comment by jerrypeng** [comment]
> The point here for a given POJO, Pulsar should maintain a basic contract: the schema generated by > Pulsar AVRO should be same/compatible with what other tools generated. Pulsar AVRO should not > > automatically attach AllowNull for all fields, otherwise we shouldn't call it AVRO schema, we should > > be calling it AlwaysAllowNullAvroSchema.

@sijie the contract you are creating when generating an avro schema using the ReflectData api is with the java class itself not some other tool or system.  If you want to provide a custom schema, we should have the way to do so.  These issues are connected.

Like I mentioned before, a user can create a Avro schema file with arbitrary fields set to null.  There is not way to get such information from a java class along.  Thus, the api currently does not suffice for general use cases when users are bring in generic avro schemas irregardless of whether allowNull is set by default.

To summarize, we can NOT set allowNull by default but it hardly solves the underlying issue.

**Comment by sijie** [comment]
> Given a POJO generated by Avro, there is no way to determine whether this POJO was generated with a schema that allowed or not for null types.
> but that's separate from the point that that alone won't guarantee we can generate the correct schema starting from the generated POJO.

@merlimat 

The issue I am creating here is for `AllowNull`. We found `AllowNull` is a problem from 2 use cases: 1) the one reported by @skyrocknroll  2) the other use case that @codelipenghui hit.  

The whole picture of @skyrocknroll `s problem is  avro file => avro generated pojo. a schema generated by avro file is not compatible by scheme generated by pulsar.AvroSchema(generated pojo). One of the problems is `AllowNull` completely changes the schema definitions. Whether removing `AllowNull` can address this problem or not is a separate issue to be address. Although I would expect Avro can handle this well. We shouldn't couple the discussion of this issue with a broad issue introduced by `AllowNull`. 

> It's a 100% correct solution for that case. I don't see what's limited about that.

@merlimat 

Generated POJO is the use case reported and discussed at slack channel. If you just handle generated POJO by using `getClassSchema`, you are not covering many other data sources which generate AVRO schema using ReflectData.

> the contract you are creating when generating an avro schema using the ReflectData api is with the java class itself not some other tool or system. 

@jerrypeng 

I am not creating any contract. They are from real use cases. Also the whole discussion is around POJO only, it is not even a cross-language issue or any user customized schema issue. (it was found when being used in cross-language)

1) if a user uses an AVRO schema file to generate a schema using avro tools (that's call this schema A), and generate a POJO class. then the user use the POJO class and use pulsar avro schema to generate another schema B. Ideally A and B should be compatible.

```
schema A => generated pojo => (pulsar AvroSchema) => schema B

                       ||
                       \/

schema A => generated pojo => ReflectData.AllowNull.parse => schema B
```

`AllowNull` is the problem to prevent them being compatible. I don't know if removing `AllowNull` can fully address this problem or not. That's a specific issue to address for Avro generated POJO. I hope Avro provides the right tools to convert back-and-forth between schema and pojos, otherwise IMO it is a problem of Avro.

If we removed `AllowNull`, the flow will be changed to following. The scope of the problem is different - whether `ReflectData` is the right tool to handle Avro generated POJO or does Avro even provide tools to guarantee such conversations.

```
schema A => generated pojo => ReflectData.parse => schema B
```

2)  Image I have a data system A (e.g. Spark or Flink) and Pulsar. I have a POJO class (e.g. UserProfile) defined across the whole organization. The schema generated in different systems are completely not compatible even they are using same POJO. When data is flowing between pulsar and other systems, the data might not be processed properly due to incompatible schema. Pulsar is the message bus for exchanging data between other system. If it produces an incompatible schema than other systems, IMO that's a very serious bug.

**Comment by congbobo184** [comment]
#3752

**Comment by jerrypeng** [comment]
@sijie that not going to work.  If you take look at the java code generated by avro tools with a given schema file its going to be not what you expect.  And from that generated class there is no way for reflect data avro library to generate the same schema unless we do the trick @merlimat mentioned

**Comment by sijie** [comment]
@jerrypeng 

> And from that generated class there is no way for reflect data avro library to generate the same schema unless we do the trick @merlimat mentioned

we can do the trick for what @merlimat mentioned for addressing the specific problem of avro generated code. but that's the separate problem of AllowNull.

**Comment by jerrypeng** [comment]
@sijie actually this does work:

schema A => generated pojo => ReflectData.parse => schema B

Schema A == . Schema B

**Comment by sijie** [comment]
@jerrypeng 

> Using the reflect data parser in avro will produce the same schema.

Glad we are on the same page now. 

> I agree we should remove AllowNull to be set by default, but we need a migration plan

I don't think we need a migration plan. We just need a flag to tell whether `AllowNull` is used or not. Explained in #3752 . Migration is a nightmare. I don't want to touch it.

**Comment by skyrocknroll** [comment]
We are using avro because avro being an standard and guaranteed to work across different part of the  whole data pipeline. Any system we use in the pipeline we will just check does it adhere to avro standards. To add more clarity , We force every engineer to first create an avro schema and get it reviewed by data platform team. We generate pojo for  java , Python and go struts. Any system violating the avro standards will breaks the system.

**Comment by merlimat** [comment]
@skyrocknroll I guess, for Python it might make sense to also allow to pass the JSON avro definition

**Comment by skyrocknroll** [comment]
@merlimat Yes, that would be awesome!

**Comment by jerrypeng** [comment]
@sijie @skyrocknroll I just experimented a little.

Given the schema that @skyrocknroll posted on the pulsar chat as an example:

```
{
  "namespace": "example.avro",
  "type": "record",
  "name": "User",
  "fields": [
    {
      "name": "name",
      "type": "string"
    },
    {
      "name": "favorite_number",
      "type": [
        "int",
        "null"
      ]
    },
    {
      "name": "favorite_color",
      "type": "string"
    },
    {
      "name": "age",
      "type": "int",
      "default": 19
    }
  ]
}
```

I did some tests using the generated java class with the ReflectData schema generator:

```
log.info("{}", new String(AvroSchema.of(User.class).getSchemaInfo().getSchema()));

log.info("{}", ReflectData.get().getSchema(User.class));

log.info("{}", ReflectData.AllowNull.get().getSchema(User.class));
```

They all produced the same result:

```
21:07:27.741 [main:org.company.Test4@44] INFO  org.company.Test4 - {"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}

21:07:27.746 [main:org.company.Test4@46] INFO  org.company.Test4 - {"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}

21:07:27.746 [main:org.company.Test4@48] INFO  org.company.Test4 - {"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}
```

regardless of  whether AllowNull is set or not.  So my question still holds what exactly does removing AllowNull change?

**Comment by jerrypeng** [comment]
Here is actually the line of code that actually just gets the "SCHEMA" field regardless of the java class:
https://github.com/apache/avro/blob/release-1.8.2/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java#L277

**Comment by sijie** [comment]
@jerrypeng 

I am surprised to see `{"name":"favorite_color","type":"string"}`is not a union type with `null` after using AllowNull.

so my question is how do you do the test? Is the User any kind of POJO class, or avro generated POJO class?

**Comment by jerrypeng** [comment]
@sijie The User class is an Avro generated class based on the schema file.

```
/**
 * Autogenerated by Avro
 *
 * DO NOT EDIT DIRECTLY
 */
package example.avro;

import org.apache.avro.specific.SpecificData;
import org.apache.avro.message.BinaryMessageEncoder;
import org.apache.avro.message.BinaryMessageDecoder;
import org.apache.avro.message.SchemaStore;

@SuppressWarnings("all")
@org.apache.avro.specific.AvroGenerated
public class User extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  private static final long serialVersionUID = -7831158737375082161L;
  public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"User\",\"namespace\":\"example.avro\",\"fields\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"favorite_number\",\"type\":[\"int\",\"null\"]},{\"name\":\"favorite_color\",\"type\":\"string\"},{\"name\":\"age\",\"type\":\"int\",\"default\":19}]}");
  public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }

  private static SpecificData MODEL$ = new SpecificData();

  private static final BinaryMessageEncoder<User> ENCODER =
      new BinaryMessageEncoder<User>(MODEL$, SCHEMA$);

  private static final BinaryMessageDecoder<User> DECODER =
      new BinaryMessageDecoder<User>(MODEL$, SCHEMA$);

  /**
   * Return the BinaryMessageDecoder instance used by this class.
   */
  public static BinaryMessageDecoder<User> getDecoder() {
    return DECODER;
  }

  /**
   * Create a new BinaryMessageDecoder instance for this class that uses the specified {@link SchemaStore}.
   * @param resolver a {@link SchemaStore} used to find schemas by fingerprint
   */
  public static BinaryMessageDecoder<User> createDecoder(SchemaStore resolver) {
    return new BinaryMessageDecoder<User>(MODEL$, SCHEMA$, resolver);
  }

  /** Serializes this User to a ByteBuffer. */
  public java.nio.ByteBuffer toByteBuffer() throws java.io.IOException {
    return ENCODER.encode(this);
  }

  /** Deserializes a User from a ByteBuffer. */
  public static User fromByteBuffer(
      java.nio.ByteBuffer b) throws java.io.IOException {
    return DECODER.decode(b);
  }

  @Deprecated public java.lang.CharSequence name;
  @Deprecated public java.lang.Integer favorite_number;
  @Deprecated public java.lang.CharSequence favorite_color;
  @Deprecated public int age;

  /**
   * Default constructor.  Note that this does not initialize fields
   * to their default values from the schema.  If that is desired then
   * one should use <code>newBuilder()</code>.
   */
  public User() {}

  /**
   * All-args constructor.
   * @param name The new value for name
   * @param favorite_number The new value for favorite_number
   * @param favorite_color The new value for favorite_color
   * @param age The new value for age
   */
  public User(java.lang.CharSequence name, java.lang.Integer favorite_number, java.lang.CharSequence favorite_color, java.lang.Integer age) {
    this.name = name;
    this.favorite_number = favorite_number;
    this.favorite_color = favorite_color;
    this.age = age;
  }

  public org.apache.avro.Schema getSchema() { return SCHEMA$; }
  // Used by DatumWriter.  Applications should not call.
  public java.lang.Object get(int field$) {
    switch (field$) {
    case 0: return name;
    case 1: return favorite_number;
    case 2: return favorite_color;
    case 3: return age;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  // Used by DatumReader.  Applications should not call.
  @SuppressWarnings(value="unchecked")
  public void put(int field$, java.lang.Object value$) {
    switch (field$) {
    case 0: name = (java.lang.CharSequence)value$; break;
    case 1: favorite_number = (java.lang.Integer)value$; break;
    case 2: favorite_color = (java.lang.CharSequence)value$; break;
    case 3: age = (java.lang.Integer)value$; break;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  /**
   * Gets the value of the 'name' field.
   * @return The value of the 'name' field.
   */
  public java.lang.CharSequence getName() {
    return name;
  }

  /**
   * Sets the value of the 'name' field.
   * @param value the value to set.
   */
  public void setName(java.lang.CharSequence value) {
    this.name = value;
  }

  /**
   * Gets the value of the 'favorite_number' field.
   * @return The value of the 'favorite_number' field.
   */
  public java.lang.Integer getFavoriteNumber() {
    return favorite_number;
  }

  /**
   * Sets the value of the 'favorite_number' field.
   * @param value the value to set.
   */
  public void setFavoriteNumber(java.lang.Integer value) {
    this.favorite_number = value;
  }

  /**
   * Gets the value of the 'favorite_color' field.
   * @return The value of the 'favorite_color' field.
   */
  public java.lang.CharSequence getFavoriteColor() {
    return favorite_color;
  }

  /**
   * Sets the value of the 'favorite_color' field.
   * @param value the value to set.
   */
  public void setFavoriteColor(java.lang.CharSequence value) {
    this.favorite_color = value;
  }

  /**
   * Gets the value of the 'age' field.
   * @return The value of the 'age' field.
   */
  public java.lang.Integer getAge() {
    return age;
  }

  /**
   * Sets the value of the 'age' field.
   * @param value the value to set.
   */
  public void setAge(java.lang.Integer value) {
    this.age = value;
  }

  /**
   * Creates a new User RecordBuilder.
   * @return A new User RecordBuilder
   */
  public static example.avro.User.Builder newBuilder() {
    return new example.avro.User.Builder();
  }

  /**
   * Creates a new User RecordBuilder by copying an existing Builder.
   * @param other The existing builder to copy.
   * @return A new User RecordBuilder
   */
  public static example.avro.User.Builder newBuilder(example.avro.User.Builder other) {
    return new example.avro.User.Builder(other);
  }

  /**
   * Creates a new User RecordBuilder by copying an existing User instance.
   * @param other The existing instance to copy.
   * @return A new User RecordBuilder
   */
  public static example.avro.User.Builder newBuilder(example.avro.User other) {
    return new example.avro.User.Builder(other);
  }

  /**
   * RecordBuilder for User instances.
   */
  public static class Builder extends org.apache.avro.specific.SpecificRecordBuilderBase<User>
    implements org.apache.avro.data.RecordBuilder<User> {

    private java.lang.CharSequence name;
    private java.lang.Integer favorite_number;
    private java.lang.CharSequence favorite_color;
    private int age;

    /** Creates a new Builder */
    private Builder() {
      super(SCHEMA$);
    }

    /**
     * Creates a Builder by copying an existing Builder.
     * @param other The existing Builder to copy.
     */
    private Builder(example.avro.User.Builder other) {
      super(other);
      if (isValidValue(fields()[0], other.name)) {
        this.name = data().deepCopy(fields()[0].schema(), other.name);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.favorite_number)) {
        this.favorite_number = data().deepCopy(fields()[1].schema(), other.favorite_number);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.favorite_color)) {
        this.favorite_color = data().deepCopy(fields()[2].schema(), other.favorite_color);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.age)) {
        this.age = data().deepCopy(fields()[3].schema(), other.age);
        fieldSetFlags()[3] = true;
      }
    }

    /**
     * Creates a Builder by copying an existing User instance
     * @param other The existing instance to copy.
     */
    private Builder(example.avro.User other) {
            super(SCHEMA$);
      if (isValidValue(fields()[0], other.name)) {
        this.name = data().deepCopy(fields()[0].schema(), other.name);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.favorite_number)) {
        this.favorite_number = data().deepCopy(fields()[1].schema(), other.favorite_number);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.favorite_color)) {
        this.favorite_color = data().deepCopy(fields()[2].schema(), other.favorite_color);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.age)) {
        this.age = data().deepCopy(fields()[3].schema(), other.age);
        fieldSetFlags()[3] = true;
      }
    }

    /**
      * Gets the value of the 'name' field.
      * @return The value.
      */
    public java.lang.CharSequence getName() {
      return name;
    }

    /**
      * Sets the value of the 'name' field.
      * @param value The value of 'name'.
      * @return This builder.
      */
    public example.avro.User.Builder setName(java.lang.CharSequence value) {
      validate(fields()[0], value);
      this.name = value;
      fieldSetFlags()[0] = true;
      return this;
    }

    /**
      * Checks whether the 'name' field has been set.
      * @return True if the 'name' field has been set, false otherwise.
      */
    public boolean hasName() {
      return fieldSetFlags()[0];
    }


    /**
      * Clears the value of the 'name' field.
      * @return This builder.
      */
    public example.avro.User.Builder clearName() {
      name = null;
      fieldSetFlags()[0] = false;
      return this;
    }

    /**
      * Gets the value of the 'favorite_number' field.
      * @return The value.
      */
    public java.lang.Integer getFavoriteNumber() {
      return favorite_number;
    }

    /**
      * Sets the value of the 'favorite_number' field.
      * @param value The value of 'favorite_number'.
      * @return This builder.
      */
    public example.avro.User.Builder setFavoriteNumber(java.lang.Integer value) {
      validate(fields()[1], value);
      this.favorite_number = value;
      fieldSetFlags()[1] = true;
      return this;
    }

    /**
      * Checks whether the 'favorite_number' field has been set.
      * @return True if the 'favorite_number' field has been set, false otherwise.
      */
    public boolean hasFavoriteNumber() {
      return fieldSetFlags()[1];
    }


    /**
      * Clears the value of the 'favorite_number' field.
      * @return This builder.
      */
    public example.avro.User.Builder clearFavoriteNumber() {
      favorite_number = null;
      fieldSetFlags()[1] = false;
      return this;
    }

    /**
      * Gets the value of the 'favorite_color' field.
      * @return The value.
      */
    public java.lang.CharSequence getFavoriteColor() {
      return favorite_color;
    }

    /**
      * Sets the value of the 'favorite_color' field.
      * @param value The value of 'favorite_color'.
      * @return This builder.
      */
    public example.avro.User.Builder setFavoriteColor(java.lang.CharSequence value) {
      validate(fields()[2], value);
      this.favorite_color = value;
      fieldSetFlags()[2] = true;
      return this;
    }

    /**
      * Checks whether the 'favorite_color' field has been set.
      * @return True if the 'favorite_color' field has been set, false otherwise.
      */
    public boolean hasFavoriteColor() {
      return fieldSetFlags()[2];
    }


    /**
      * Clears the value of the 'favorite_color' field.
      * @return This builder.
      */
    public example.avro.User.Builder clearFavoriteColor() {
      favorite_color = null;
      fieldSetFlags()[2] = false;
      return this;
    }

    /**
      * Gets the value of the 'age' field.
      * @return The value.
      */
    public java.lang.Integer getAge() {
      return age;
    }

    /**
      * Sets the value of the 'age' field.
      * @param value The value of 'age'.
      * @return This builder.
      */
    public example.avro.User.Builder setAge(int value) {
      validate(fields()[3], value);
      this.age = value;
      fieldSetFlags()[3] = true;
      return this;
    }

    /**
      * Checks whether the 'age' field has been set.
      * @return True if the 'age' field has been set, false otherwise.
      */
    public boolean hasAge() {
      return fieldSetFlags()[3];
    }


    /**
      * Clears the value of the 'age' field.
      * @return This builder.
      */
    public example.avro.User.Builder clearAge() {
      fieldSetFlags()[3] = false;
      return this;
    }

    @Override
    @SuppressWarnings("unchecked")
    public User build() {
      try {
        User record = new User();
        record.name = fieldSetFlags()[0] ? this.name : (java.lang.CharSequence) defaultValue(fields()[0]);
        record.favorite_number = fieldSetFlags()[1] ? this.favorite_number : (java.lang.Integer) defaultValue(fields()[1]);
        record.favorite_color = fieldSetFlags()[2] ? this.favorite_color : (java.lang.CharSequence) defaultValue(fields()[2]);
        record.age = fieldSetFlags()[3] ? this.age : (java.lang.Integer) defaultValue(fields()[3]);
        return record;
      } catch (java.lang.Exception e) {
        throw new org.apache.avro.AvroRuntimeException(e);
      }
    }
  }

  @SuppressWarnings("unchecked")
  private static final org.apache.avro.io.DatumWriter<User>
    WRITER$ = (org.apache.avro.io.DatumWriter<User>)MODEL$.createDatumWriter(SCHEMA$);

  @Override public void writeExternal(java.io.ObjectOutput out)
    throws java.io.IOException {
    WRITER$.write(this, SpecificData.getEncoder(out));
  }

  @SuppressWarnings("unchecked")
  private static final org.apache.avro.io.DatumReader<User>
    READER$ = (org.apache.avro.io.DatumReader<User>)MODEL$.createDatumReader(SCHEMA$);

  @Override public void readExternal(java.io.ObjectInput in)
    throws java.io.IOException {
    READER$.read(this, SpecificData.getDecoder(in));
  }

}
```

**Comment by jerrypeng** [comment]
@sijie That is the workflow you were worried about right? i.e.
```
schema A => generated pojo => ReflectData.parse => schema B
```
and schema A should equal schema B

**Comment by sijie** [comment]
@jerrypeng I got a different result than yours

```
Pulsar AllowNull     :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":["null","string"],"default":null},{"name":"favorite_number","type":["null","int"],"default":null},{"name":"favorite_color","type":["null","string"],"default":null},{"name":"age","type":"int"}]}
Avro Not AllowNull   :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}
Avro AllowNull       :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}
```

**Comment by jerrypeng** [comment]
@sijie I think the reason why ```AvroSchema.of(User.class)``` is producing a different result is because we shade all the Avro dependencies and that is interfering it from detecting that its a generated class, but  ```Avro Not AllowNull``` is the same as ```Avro Not AllowNull``` which we are using the exactly the later in AvroSchema.  Maybe we should figure out a way to getting the shading to work with Avro?

**Comment by jerrypeng** [comment]
Here is a Test project i have created:

https://github.com/jerrypeng/TestAvro

I verified the reason why the AvroSchema part of the shaded pulsar-client library outputs a different schema is because of shading

**Comment by sijie** [comment]
@jerrypeng  correct. I added one more to complete the picture.

```
        log.info("Pulsar AllowNull                  :{}", new String(AvroSchema.of(User.class).getSchemaInfo().getSchema()));
        log.info("Pulsar shaded AVRO NOT AllowNull  :{}", org.apache.pulsar.shade.org.apache.avro.reflect.ReflectData.get().getSchema(User.class));
        log.info("Avro AllowNull                    :{}", ReflectData.AllowNull.get().getSchema(User.class));
        log.info("Avro Not AllowNull                :{}", ReflectData.get().getSchema(User.class));
```

```
Pulsar AllowNull                  :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":["null","string"],"default":null},{"name":"favorite_number","type":["null","int"],"default":null},{"name":"favorite_color","type":["null","string"],"default":null},{"name":"age","type":"int"}]}
Pulsar shaded AVRO NOT AllowNull  :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":"int"},{"name":"favorite_color","type":"string"},{"name":"age","type":"int"}]}
Avro AllowNull                    :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}
Avro Not AllowNull                :{"type":"record","name":"User","namespace":"example.avro","fields":[{"name":"name","type":"string"},{"name":"favorite_number","type":["int","null"]},{"name":"favorite_color","type":"string"},{"name":"age","type":"int","default":19}]}
```

As you described above, AllowNull doesn't matter if the POJO is generated by AVRO, because AVRO will try to use `SCHEMA$` field to retrieve the schema. But If the POJO is not generated by AVRO, AVRO will attempt to parse it and AllowNull will generating two different incompatible schemas.

my tests avoid covered all the cases.

1) `Avro AllowNull` and `Avro Not AllowNull` are using unshaded avro library to parse avro generated POJO. in this way, they will always produce the same schema no matter `AllowNull` used or not. 

2) `Pulsar AllowNull` and `Pulsar shade avro Not Allow Null` are using shaded avro library to parse avro generated POJO. In this way, they will produce a completely different schema. `AllowNull` matters here. 

The 2) indicates two problems:

a) for normal POJO (not generated by AVRO), `AllowNull` is producing incompatible schemas. so we need #3752 

b) for avro generated POJO, shaded AVRO library (the one used by Pulsar) doesn't recognize it is an AVRO generated POJO. We need a separate fix to make sure Pulsar shaded AVRO working with avro generated POJO.

**Comment by jerrypeng** [comment]
@sijie yup though I am not sure setting AllowNull is producing an incompatible schema.  Incompatible to what?  I guess if a user wrote some code:

```
ReflectData.get().getSchema(MyClass.class)
```

and they didn't set allow null then it wouldn't be compatible with the current schema generated from AvroSchema but my implementation could just have easily set allow null.  Thus, this doesn't seem like big deal.  I think the bigger problem here is the one with shading and how generated java classes are handled since that is a API/standard that should be preserved

**Comment by sijie** [comment]
> yup though I am not sure setting AllowNull is producing an incompatible schema. Incompatible to what? 

@jerrypeng `ReflectData.AllowNull` and `ReflectData` are generating two different incompatible schemas. I have been explaining to you guys in previous comments: my broader concern is NOT about generated POJO, is about using a same POJO class across many different systems using AVRO as the standard serialization format where people are using `ReflectData` not `ReflectData.AllowNull` (the second case explained at https://github.com/apache/pulsar/issues/3741#issuecomment-469506795). 

> my implementation could just have easily set allow null. Thus, this doesn't seem like big deal and arbitrary.

well it is a big deal when exchanging data between systems using same POJO.

> Should we not preserve that semantic? 

If Pulsar uses AVRO on POJOs, I don't think it is the responsibility for Pulsar to preserve its own semantic. It should preserve what AVRO *preserves* for everyone because this is the rules of all other systems follow when using AVRO. That's the most important thing when AVRO is used as a standard data format across data systems. Currently Pulsar's AllowNull implementation violates it.

> but I think the bigger problem here is the one with shading 

The shading bug is new to me. I agreed that we need to fix it.

**Comment by skyrocknroll** [comment]
Avro Schema compatability https://docs.confluent.io/current/schema-registry/docs/avro.html#compatibility-types
